clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/ProducerTracker.java

```diff
@@ -105,49 +103,58 @@ public void setPartitionState(int partition, ProducerPartitionState state) {
   }
 
   private void setPartitionState(int partition, Segment segment) {
-    ReentrantLock partitionLock = getPartitionLock(partition);
-    partitionLock.lock();
+    ReentrantLockWithRef<Segment> lockAndSegment = getLockAndSegment(partition);
+    lockAndSegment.lock();
     try {
-      if (segments.containsKey(partition)) {
+      if (lockAndSegment.containsRef()) {
         logger.info(
             "{} will overwrite previous state for partition: {}, Previous state: {}, New state: {}",
             this,
             partition,
-            segments.get(partition),
+            lockAndSegment.getRef(),
             segment);
       } else {
         logger.info("{} will set state for partition: {}, New state: {}", this, partition, segment);
       }
-      segments.put(partition, segment);
+      lockAndSegment.setRef(segment);
     } finally {
-      partitionLock.unlock();
+      lockAndSegment.unlock();
     }
   }
 
   public void cloneProducerStates(int partition, ProducerTracker destProducerTracker) {
-    if (!segments.containsKey(partition)) {
-      // This producer didn't write anything to requested partition
+    ReentrantLockWithRef<Segment> lockAndSegment = getLockAndSegment(partition);
+    if (!lockAndSegment.containsRef()) {
+      // This producer didn't write anything to requested partition (lock-free check)
       return;
     }
-    ReentrantLock partitionLock = getPartitionLock(partition);
-    partitionLock.lock();
+    lockAndSegment.lock();
     try {
-      Segment sourceSegment = segments.get(partition);
+      if (!lockAndSegment.containsRef()) {
+        // This producer didn't write anything to requested partition
+        return;
+      }
```

Comment:
Why do we need this check inside the lock scope? Any race condition this logic is trying to guard? If yes, can you leave some javadoc?
Same for the following function.

Reply:
It is possible that another thread calls `removeRef` between the first (lock-free) check and the lock acquisition.

But if we think this looks too complicated, then I could remove the pre-lock check.

Reply:
I guess that is fine to leave it in this way.